framework/test/gtest/BundleValidationTest.cpp

```diff
@@ -0,0 +1,141 @@
+  // Framework not to start the bundle and it should not be loaded
+  // into the process.
+  ASSERT_EQ(bundleA.GetState(),
+            cppmicroservices::Bundle::State::STATE_RESOLVED);
+
+  f.Stop();
+  f.WaitForStop(std::chrono::milliseconds::zero());
+}
+
+TEST(BundleValidationTest, BundleValidationSuccess) {
+  using validationFuncType = std::function<bool(const cppmicroservices::Bundle&)>;
+
+  validationFuncType validationFunc = [](const cppmicroservices::Bundle&) -> bool {
+    return true;
+  };
+  cppmicroservices::FrameworkConfiguration configuration{
+    { cppmicroservices::Constants::FRAMEWORK_BUNDLE_VALIDATION_FUNC,
+      validationFunc }
+  };
+
+  auto f =
+    cppmicroservices::FrameworkFactory().NewFramework(std::move(configuration));
+  ASSERT_NO_THROW(f.Start());
+
+  auto bundleA =
+    cppmicroservices::testing::InstallLib(f.GetBundleContext(), "TestBundleA");
+  ASSERT_TRUE(bundleA);
+
+  ASSERT_NO_THROW(bundleA.Start());
+  // a bundle validation function which returns true must cause the
+  // Framework to start the bundle and it should be loaded
+  // into the process.
+  ASSERT_EQ(bundleA.GetState(),
+            cppmicroservices::Bundle::State::STATE_ACTIVE);
+
+  f.Stop();
+  f.WaitForStop(std::chrono::milliseconds::zero());
+}
+
+TEST(BundleValidationTest, BundleValidationFunctionException) {
+  using validationFuncType = std::function<bool(const cppmicroservices::Bundle&)>;
+
+  validationFuncType validationFunc = [](const cppmicroservices::Bundle&) -> bool {
+    throw std::runtime_error("foobar");
+  };
+  cppmicroservices::FrameworkConfiguration configuration{
+    { cppmicroservices::Constants::FRAMEWORK_BUNDLE_VALIDATION_FUNC,
+      validationFunc }
+  };
+
+  auto f =
+    cppmicroservices::FrameworkFactory().NewFramework(std::move(configuration));
+  ASSERT_NO_THROW(f.Start());
+
+  bool receivedBundleValidationErrorEvent{ false };
+  auto token = f.GetBundleContext().AddFrameworkListener(
+    [&receivedBundleValidationErrorEvent](
+      const cppmicroservices::FrameworkEvent& evt) {
+      if (evt.GetType() ==
+          cppmicroservices::FrameworkEvent::Type::FRAMEWORK_WARNING &&
+          evt.GetBundle().GetSymbolicName() == "TestBundleA") {
+        receivedBundleValidationErrorEvent = true;
+      }
+    });
+
+  auto bundleA =
+    cppmicroservices::testing::InstallLib(f.GetBundleContext(), "TestBundleA");
+  ASSERT_TRUE(bundleA);
+
+  ASSERT_THROW(bundleA.Start(), cppmicroservices::SecurityException);
+  // a bundle validation function which returns false must cause the
+  // Framework not to start the bundle and it should not be loaded
+  // into the process.
+  ASSERT_EQ(bundleA.GetState(),
+            cppmicroservices::Bundle::State::STATE_RESOLVED);
+  ASSERT_TRUE(receivedBundleValidationErrorEvent);
+
+  f.GetBundleContext().RemoveListener(std::move(token));
+  f.Stop();
+  f.WaitForStop(std::chrono::milliseconds::zero());
+}
+
+#endif
```

Comment:
Same comment here as for the DS tests. Maybe a concurrent GetService test for failures and exceptions would be useful?